packages/drawer/src/ux-modal-service.ts

```diff
@@ -0,0 +1,277 @@
+    if (options.view && bindingContext.composeViewModel) {
+      bindingContext.composeViewModel.view = options.view;
+    }
+    if (options.model && bindingContext.composeViewModel) {
+      bindingContext.composeViewModel.model = options.model;
+    }
+
+    const controllers = (childView as any).controllers as Controller[];
+    
+    const drawer: UxDrawer =  controllers[0].viewModel as UxDrawer;
+    const drawerIndex = this.drawerIndex;
+    const whenClosed: Promise<ModalServiceResult> = new Promise((resolve) => {
+      this.eventAggregator.subscribeOnce(`drawer-${drawerIndex}-resolve`, (result: ModalServiceResult) => {
+        resolve(result);
+      });      
+    });
+
+    (drawer as UxDrawer & ModalServiceDrawer).whenClosed = () => {
+      return whenClosed;
+    }
+    bindingContext.dismiss = () => {
+      drawer.detached();
+      const parent = drawer.element.parentNode;
+      if (!parent) { return; }
+      parent.removeChild(drawer.element);
+      this.eventAggregator.publish(`drawer-${drawerIndex}-resolve`, {
+        wasCancelled: true,
+        output: null
+      });
+    }
+    bindingContext.ok = (event: CustomEvent) => {
+      drawer.detached();
+      const parent = drawer.element.parentNode;
+      if (!parent) { return; }
+      parent.removeChild(drawer.element);
+      this.eventAggregator.publish(`drawer-${drawerIndex}-resolve`, {
+        wasCancelled: false,
+        output: event.detail
+      });
+    }
+    return (drawer as UxDrawer & ModalServiceDrawer);
+  }
+
+  public async callCanDeactivate(layer: ModalLayer, result: ModalServiceResult): Promise<boolean> {
+    if (layer.bindingContext && layer.bindingContext.composeViewModel && layer.bindingContext.composeViewModel.currentViewModel) {
+      const vm = layer.bindingContext.composeViewModel.currentViewModel;
+      if (typeof vm.canDeactivate === 'function') {
+        try {
+          const can = (await vm.canDeactivate.call(vm, result) === false) ? false : true;
+          return can;
+        } catch (error) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  public async callDetached(layer: ModalLayer): Promise<void> {
+    if (layer.bindingContext && layer.bindingContext.composeViewModel && layer.bindingContext.composeViewModel.currentViewModel) {
+      const vm = layer.bindingContext.composeViewModel.currentViewModel;
+      if (typeof vm.detached === 'function') {
+        await vm.detached.call(vm);
+      }
+    }
+    return;
+  }
+
+  public async callDeactivate(layer: ModalLayer, result: ModalServiceResult): Promise<void> {
+    if (layer.bindingContext && layer.bindingContext.composeViewModel && layer.bindingContext.composeViewModel.currentViewModel) {
+      const vm = layer.bindingContext.composeViewModel.currentViewModel;
+      if (typeof vm.deactivate === 'function') {
+        await vm.deactivate.call(vm, result);
+      }
+    }
+    return;
+  }
+
+  public async cancel(drawer?: UxDrawer) {
```

Comment:
does this need to be `async`?

Reply:
I forgot the `await` in the `layer.drawer.dismiss()` line.